Automattic/woocommerce-payments
# IPP Styled Receipts: Print receipt

Fixes #3021 and #3224

#### Changes proposed in this Pull Request




This PR exposes a new REST API endpoint to generate an IPP receipt for printing. The new endpoint will collect all the data needed to generate the receipt, and will return an HTML version of it. The client can then use the returned HTML to print the receipt.

We did some tests using a thermal printer to validate the implementation (thanks again @malinajirka ). After some tweaks, this is how the receipt looks like:

![20211125_185940](https://user-images.githubusercontent.com/1553182/144287188-790bad55-93ad-4920-a7e7-12238b87b81e.jpg)

I added the new endpoint to the payment readers controller because the receipts are strictly related to the readers. A client can do a `GET` request to `payments/readers/receipts/payment_id ` to fetch the receipt. 

Initially, we had the idea of caching the receipt, to get some performance savings. But during development we decided to change the approach. The generation of the receipt in HTML does not seem to be a costly operation and it will be printed just once most of the times. With these things in mind, the complexity of caching the receipt in the database seemed to me not necessary at the moment.

 





#### Testing instructions





To test this feature in your local environment you will need:

- Connect your local WC Pay client and server.
- Create an order with Cash on Delivery as payment method. Please use the proxied url of your local site.
- Start listening for webhooks in your local WC Pay server. You'll need this to refresh your order status after next step.
- Use the WooCommerce Mobile App to pay the order. You can follow this great guide to set it up locally: pdjVC1-9M-p2
- Once the payment is suceeded, you'll need to note the payment id reference i.e: you can find it in the order details
- Using Postman (note that you'll need to set up Authentication as explained in this guide: PbTz5e-12J-p2) make a GET request to `/wp-json/wc/v3/payments/readers/receipts/<payment_id>`.
- You should be able to check the HTML response within Postman. Hint: There is a preview button that let's you see the rendered HTML

![Screen Shot 2021-12-01 at 20 07 19](https://user-images.githubusercontent.com/1553182/144297787-ef02f4ed-352f-402b-8e78-1b9ecae93f1a.png)
-------------------


**Post merge**
